craft_store/request_client.py

```diff
@@ -0,0 +1,257 @@
+        response_json: dict[str, Any] = self.response.json()
+        try:
+            # Charmhub uses error-list.
+            error_list: list[dict[str, str]] = response_json["error-list"]
+        except KeyError:
+            # Snap Store uses error_list.
+            error_list = response_json["error_list"]
+
+        return error_list
+
+
+class StoreRequestClient:
+    """Client which talks with the Store through API."""
+
+    def __init__(
+        self,
+        *,
+        auth: httpx.Auth,
+        timeout: float = 30.0,
+    ) -> None:
+        self._client = httpx.Client(auth=auth, timeout=timeout)
+
+    def get(
+        self,
+        url: str,
+        **kwargs: Any,  # noqa: ANN401
+    ) -> httpx.Response:
+        """Sent GET request to the Store."""
+        return self.request("GET", url, **kwargs)
+
+    def post(
+        self,
+        url: str,
+        **kwargs: Any,  # noqa: ANN401
+    ) -> httpx.Response:
+        """Sent POST request to the Store."""
+        return self.request("POST", url, **kwargs)
+
+    def put(
+        self,
+        url: str,
+        **kwargs: Any,  # noqa: ANN401
+    ) -> httpx.Response:
+        """Sent DELETE request to the Store."""
+        return self.request("PUT", url, **kwargs)
+
+    def delete(
+        self,
+        url: str,
+        **kwargs: Any,  # noqa: ANN401
+    ) -> httpx.Response:
+        """Sent DELETE request to the Store."""
+        return self.request("DELETE", url, **kwargs)
+
+    def request(
+        self,
+        http_method: str,
+        url: str,
+        *,
+        headers: dict[str, str] | None = None,
+        check: bool = True,
+        check_store_task_status: bool = False,
+        **kwargs: Any,  # noqa: ANN401
+    ) -> httpx.Response:
+        """Issue HTTP request against given URL."""
+        if headers is None:
+            headers = {}
+        try:
+            logger.debug("Calling method %s on endpoint %s", http_method, url)
+            logger.debug("with query parameters: %s", kwargs.get("params"))
+            logger.debug("with data: %s", kwargs.get("json"))
+            response = self._client.request(
+                http_method,
+                url,
+                headers=headers,
+                **kwargs,
+            )
+        except httpx.HTTPError as err:
+            logger.debug(
+                "Error occurred when processing request to %r",
+                err.request.url,
+            )
+            raise StoreConnectionError(
+                message="Error response received from store",
+                resolution="Check your network connectivity and try again.",
+            ) from err
+        else:
+            if check:
+                self._check_response_code(response)
+
+            # TODO: check if we want that as a part of API
+            # or handle it in service directly
+            if check_store_task_status:
+                self._check_store_task_status(response)
```

Comment:
It's store specific, should we move it outside of the scope of simple request / response client ?

Reply:
TBH I don't really understand why this class exists in its current form. HTTPX [does the same logging](https://www.python-httpx.org/logging/) and otherwise it seems to just be a wrapper around the HTTPX client.

The `_check_response_code` method is worth keeping for our base `GatewayClient`, but `_check_store_task_status` seems more of a service thing to do. `craft-store` should (mostly) just handle single requests in each method.

How I'd imagine our client would work would be more like:

```python
class PublisherGatewayClient(Client):
    def register_name(self, ...) -> str:
       endpoint = f"v1/{self._namespace}"
        request_json = ...  # generate request JSON
        response = self._client.post(endpoint, json=request_json)
        self._check_response_code(response)
        return str(response.json()["id"])
```

so we wouldn't need the various request methods.